### ISSUE TITLE
UTCDateTime oddities in obspy 1.1.0

### ISSUE THREAD

**Issue by d-chambers** [obspy/obspy#2034]
version = 1.1.0 on Ubuntu 16.04 with python 3.6 installed via Anaconda

I have run into a few surprising behaviours dealing with UTCDateTime objects that are very close in time . The issues are primarily due to rounding, floating point number weirdness, and a few potential inconsistencies in the UTCDateTime object. The result is some rather surprising behaviours. For example, check out [this gist](https://gist.github.com/d-chambers/e1469318f79a84f5cc829545a60ad218).

Although there may be several issues to address here, the behaviour I rely on that is no longer consistent is this: 
Given the default precision of 6, any datetime object should be equal to a new datetime object created using the year, month, day, hour, minute, second, and microsecond attributes. This would mean the following function should always return True (unless the precision > 6)

```python
import obspy

TIME_NAMES = ('year', 'month', 'day', 'hour', 'minute', 'second',
              'microsecond')


def can_serialize_with_time_params(arg):
    """ return True if using UTC time values (year, month, day, etc.) 
    provide enough information to reconstruct the object. """
    utc = obspy.UTCDateTime(arg)
    time_dict = {x: getattr(utc, x) for x in TIME_NAMES}
    utc2 = obspy.UTCDateTime(**time_dict)
    return utc == utc2
```
 
However, this is not the case. For example

```python
can_serialize_with_time_params(1515174511.1984465)  # False
can_serialize_with_time_params(1515174511.1984463)  # True
```

Am I wrong to assume UTCDateTime objects should behave in this way? I don't mind working on a fix but I wanted to make sure everyone who cares is on the same page.

**Comment by krischer** [comment]
Without looking at this in a lot of detail I would say you are right: This is a bug. I'll delegate it to @barsch for now who has the deepest understanding of our `UTCDateTime` objects.

Another issue with some weirdness due to limited precision in the `UTCDateTime` objects: #1766

**Comment by megies** [comment]
Two comments/observations..

---
First, since you're initializing with a POSIX timestamp, floating point accuracy hits when converting to integer nanoseconds (the new internal representation), since floating point resolution is `1e-15` and thus not reaching down to nanoseconds for the large values of order `~1e9` (notice the last two digits `...92` in the integer nanosecond representation):

```python
In [1]: x = 1515174511.1984465

In [2]: t = UTCDateTime(x)

In [3]: t._ns
Out[3]: 1515174511198446592

```
---

Second, when you're querying the microsecond value that you then use to initialize a new timestamp object, the internal integer nanoseconds are converted to integer microsecond, cutting away any decimals in the process (I guess one could argue that we should instead round up/down?):

```python
   809     def _get_microsecond(self):                                                                                                 
   810         """                                                                                                                     
   811         Returns microseconds as an integer.                                                                                     
   812                                                                                                                                 
   813         :rtype: int                                                                                                             
   814         :return: Returns microseconds as an integer.                                                                            
   815                                                                                                                                 
   816         .. rubric:: Example                                                                                                     
   817                                                                                                                                 
   818         >>> dt = UTCDateTime(2012, 2, 11, 10, 11, 12, 345234)                                                                   
   819         >>> dt.microsecond                                                                                                      
   820         345234                                                                                                                  
   821         """                                                                                                                     
   822         return int(self._ns % 10**9 // 1000)                                                                                    
```

Then, when you use that value to initialize a new timestamp (as opposed to using the original floating point POSIX timestamp), you're not running into floating point precision issues:

```python
In [4]: TIME_NAMES = ('year', 'month', 'day', 'hour', 'minute', 'second',
   ...:               'microsecond')

In [5]: time_dict = {x_: getattr(t, x_) for x_ in TIME_NAMES}

In [6]: time_dict
Out[6]: 
{'day': 5,
 'hour': 17,
 'microsecond': 198446,
 'minute': 48,
 'month': 1,
 'second': 31,
 'year': 2018}

In [7]: t2 = UTCDateTime(**time_dict)

In [8]: t.microsecond
Out[8]: 198446

In [9]: t._ns
Out[9]: 1515174511198446592

In [10]: t2._ns
Out[10]: 1515174511198446000
```

---

So, I guess the bottom line is that..
 - when initializing from POSIX timestamp, you shouldn't expect accuracy to anything below microseconds and internally the nanosecond integer is subject to floating point accuracy issues
 - string representation is rounding to microsecond while accessing the `microsecond` property gets you a `floor()` representation of the microsecond value (this might have to be changed?!)

**Comment by megies** [comment]
I'll file this under 1.1.1 for now. Feel free to change to 1.2.0 if we decide to not alter behavior in the maintenance line.

**Comment by d-chambers** [comment]
@megies thanks for the insight. I think (without looking deeply into the code) rounding rather than flooring the microseconds would probably solve most of the issues.

**Comment by barsch** [comment]
@d-chambers pls feel free to add test cases to #1766 as this should be handled IMHO in one PR

**Comment by d-chambers** [comment]
@barsch, ok will do. I should have some time tomorrow to work on it.

**Comment by d-chambers** [comment]
I have some tests written but there are still some odd corner cases here, and I am not really sure how the UTCDateTime objects should behave. Consider this one:

```python
from obspy import UTCDateTime
utc1 = UTCDateTime(1515174511.1984463)
utc2 = UTCDateTime(1515174511.1984458)
utc1 == utc2  # False
str(utc1) == str(utc2)  # True
```

The str rep in both cases is: `2018-01-05T17:48:31.198446Z`. But when we check for equality they are not equal. 
This seems correct because the objects are .5 ms apart, but also seems incorrect because the string representation is the same and yet the objects are not equal. 

The reverse is also possible:
```python 
from obspy import UTCDateTime
utc1 = UTCDateTime(1515174511.1984463)
utc2 = UTCDateTime(1515174511.1984465)
utc1 == utc2  # True
str(utc1) == str(utc2)  # False
```
It seems we have a leaky abstraction on our hands. Maybe we will just have to live with not being able to infer anything about the result of comparing two UTCDateTime objects by their strings representations?

**Comment by barsch** [comment]
string comparison and UTCDateTime object comparison should IMHO behave the same way - did you test with the patch in #1766?

**Comment by megies** [comment]
Maybe we should deprecate initializing UTCDateTime objects from POSIX timestamp in seconds (to avoid floating point inaccuracies altogether)? Or at least show a warning about floating point accuracy issues?

**Comment by krischer** [comment]
> Maybe we should deprecate initializing UTCDateTime objects from POSIX timestamp in seconds (to avoid floating point inaccuracies altogether)? Or at least show a warning about floating point accuracy issues?

I'm not in favor of that and its IMHO a valid use case. Also I feel like this would be too disruptive (I know I use this pattern a lot) and if people use floating point numbers its their responsibility to deal with the resulting accuracy issues.

**Comment by d-chambers** [comment]
closed with  #2049
